drivers/i2s/i2s_mcux_flexcomm.c

```diff
@@ -0,0 +1,897 @@
+		LOG_DBG("TX STOPPED");
+		return;
+	}
+
+	switch (strm->state) {
+	case I2S_STATE_RUNNING:
+	case I2S_STATE_STOPPING:
+		/* get the next buffer from queue */
+		ret = k_msgq_get(&strm->in_queue, &buffer, K_NO_WAIT);
+		if (ret == 0) {
+			/* config the DMA */
+			i2s_mcux_config_dma_blocks(dev, I2S_DIR_TX, (uint32_t *)buffer);
+			dma_config(strm->dev_dma, strm->channel,
+				   &strm->dma_cfg);
+			k_msgq_put(&strm->out_queue, &buffer, K_NO_WAIT);
+			dma_start(strm->dev_dma, strm->channel);
+		}
+
+		if (ret || status) {
+			/*
+			 * DMA encountered an error (status != 0)
+			 * or
+			 * No bufers in input queue
+			 */
+			LOG_DBG("DMA status %08x channel %u k_msgq_get ret %d",
+				status, channel, ret);
+			if (strm->state == I2S_STATE_STOPPING) {
+				strm->state = I2S_STATE_READY;
+			} else {
+				strm->state = I2S_STATE_ERROR;
+			}
+			i2s_mcux_tx_stream_disable(dev, false);
+		}
+		break;
+	case I2S_STATE_ERROR:
+		strm->state = I2S_STATE_READY;
+		i2s_mcux_tx_stream_disable(dev, false);
+		break;
+	}
+}
+
+static void i2s_mcux_dma_rx_callback(const struct device *dma_dev, void *arg,
+				uint32_t channel, int status)
+{
+	struct device *dev = (struct device *)arg;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *strm = &dev_data->dma_rx;
+	void *buffer;
+	int ret;
+
+	LOG_DBG("rx cb: %d", strm->state);
+
+	if (status != 0) {
+		strm->state = I2S_STATE_ERROR;
+		i2s_mcux_rx_stream_disable(dev, false);
+		return;
+	}
+
+	switch (strm->state) {
+	case I2S_STATE_STOPPING:
+	case I2S_STATE_RUNNING:
+		/* retrieve buffer from input queue */
+		ret = k_msgq_get(&strm->in_queue, &buffer, K_NO_WAIT);
+		if (ret != 0) {
+			LOG_ERR("get buffer from in_queue %p failed (%d)",
+				&strm->in_queue, ret);
+			return;
+		}
+		/* put buffer to output queue */
+		ret = k_msgq_put(&strm->out_queue, &buffer, K_NO_WAIT);
+		if (ret != 0) {
+			LOG_ERR("buffer %p -> out_queue %p err %d", buffer,
+				&strm->out_queue, ret);
+		}
+		if (strm->state == I2S_STATE_RUNNING) {
+			/* allocate new buffer for next audio frame */
+			ret = k_mem_slab_alloc(strm->cfg.mem_slab, &buffer, K_NO_WAIT);
+			if (ret != 0) {
+				LOG_ERR("buffer alloc from slab %p err %d",
+					strm->cfg.mem_slab, ret);
+				i2s_mcux_rx_stream_disable(dev, false);
+				strm->state = I2S_STATE_ERROR;
+			} else {
+				i2s_mcux_config_dma_blocks(dev, I2S_DIR_RX,
+										(uint32_t *)buffer);
+				dma_config(strm->dev_dma, strm->channel,
+					   &strm->dma_cfg);
+				dma_start(strm->dev_dma, strm->channel);
+
+				/* put buffer in input queue */
+				ret = k_msgq_put(&strm->in_queue, &buffer, K_NO_WAIT);
+				if (ret != 0) {
+					LOG_ERR("buffer %p -> in_queue %p err %d",
+					buffer, &strm->in_queue, ret);
+				}
+			}
+		} else {
+			/* Received a STOP/DRAIN trigger */
+			i2s_mcux_rx_stream_disable(dev, true);
+			strm->state = I2S_STATE_READY;
+		}
+		break;
+	case I2S_STATE_ERROR:
+		i2s_mcux_rx_stream_disable(dev, true);
+		strm->state = I2S_STATE_READY;
```

Comment:
Seems wrong. Once we get into the error state the application needs to be notified and undertake relevant steps to recover. This shouldn't happen automatically.

Reply:
Got it, I will delete these cases and let the application handle it.

Reply:
The implementation in stm32 driver looks good. In case of error the driver disables the stream but preserves the interface state as `I2S_STATE_ERROR`. In your previous implementation you've been changing the interface state to `I2S_STATE_READY`.

The reason why we shouldn't change the interface state to ready is to make sure application is aware that continuity of the data stream was broken. Samples might be missing or corrupted. Driver cannot really recover from this situation. It's application that needs to worry about it.

Reply:
Sounds good, then I will disable the stream and set state to I2S_ERROR.

Reply:
Thanks, I have updated the implementation to drop the line switching to `I2S_STATE_READY` state